### PR TITLE
Simplify `HX711.read()` logic

### PR DESCRIPTION
**Pull request by zfields**

### CODE REVIEW

**Comment by zfields** [comment]
@bogde Fixed the merge conflict. It's ready for you to test and merge. Cheers!

**Comment by FabianInostroza** [comment]
Why are you inverting bits and adding 1?
I think that that formula in not necessary here.

**Comment by bogde** [comment]
Thank you very much!

**Comment by zfields** [comment]
@FabianInostroza The [datasheet](https://cdn.sparkfun.com/datasheets/Sensors/ForceFlex/hx711_english.pdf) indicates the value is a 24-bit two's complement (signed) value

**Comment by FabianInostroza** [comment]
The C (and C++) language already encodes signed integers as two complement, so you only need to do the padding right.

**Comment by zfields** [comment]
@FabianInostroza I would agree with you 100%, except the "reference driver" in the datasheet performs this transformation on the result `Count=Count^0x800000`, which makes me believe it's actually returning the value in 2's complement.

**Comment by FabianInostroza** [comment]
I'm not saying that the value returned is not in 2's complement. I'm saying that the additional conversion you are doing is unnecessary, it only flips the sign of the result, which is wrong but not as wrong as the other problem you have in your code, the endianness.

The formula you are using for the 2's complement conversion is only useful for US to find the binary representation of a signed decimal, the computer doesn't need that formula.

I wrote some simple test, if you have time please check it https://gist.github.com/FabianInostroza/a883c35e6e6135a71cbf0d2b3536628a

**Comment by zonque** [comment]
@FabianInostroza Thanks a lot for unwrapping this. Yes indeed, the endianess in this PR is broken and cannot have worked on any 8-bit or 32-bit Arduino. I was scratching my heads over hardware bugs until I found this thread. The first byte read from the hardware has to be interpreted as the most significant one, and the memory aliasing through that union makes it the least significant on little-endian architectures.

@zfields, may I ask how you tested this?

File: HX711.cpp [hunk 1]
```diff
@@ -44,50 +44,48 @@ void HX711::set_gain(byte gain) {
...
-    byte filler = 0x00;
+    // Byte:     0        1        2        3
+    // Bits:  76543210 76543210 76543210 76543210
+    // Data: |--------|--------|--------|--------|
+    // Bit#:  33222222 22221111 11111100 00000000
+    //        10987654 32109876 54321098 76543210
+    union DataBuffer {
+        byte data[4];
+        long value;
+    } data_buffer;
```
Comment (ReDetection) [inline HX711.cpp]: I see you changed indentation style. This is a bad manner to mix different indentations.

File: HX711.cpp [hunk 2]
```diff
@@ -44,50 +44,48 @@ void HX711::set_gain(byte gain) {
...
-    data[1] = shiftIn(DOUT, PD_SCK, MSBFIRST);
-    data[0] = shiftIn(DOUT, PD_SCK, MSBFIRST);
+    // Pulse the clock pin 24 times to read the data
+    data_buffer.data[1] = shiftIn(DOUT, PD_SCK, MSBFIRST);
+    data_buffer.data[2] = shiftIn(DOUT, PD_SCK, MSBFIRST);
+    data_buffer.data[3] = shiftIn(DOUT, PD_SCK, MSBFIRST);
 
-	// set the channel and the gain factor for the next reading using the clock pin
-	for (unsigned int i = 0; i < GAIN; i++) {
+	// Set the channel and the gain factor for the next reading using the clock pin
```
Comment (ReDetection) [inline HX711.cpp]: What was the purpose of changing this comment change? How does this relates to the logic simplifying?

File: HX711.cpp [hunk 3]
```diff
@@ -44,50 +44,48 @@ void HX711::set_gain(byte gain) {
...
+	for (unsigned int i = GAIN ; 0 < i ; --i) {
```
Comment (ReDetection) [inline HX711.cpp]: What is the purpose of reversing the loop? I see no effect.

File: HX711.cpp [hunk 4]
```diff
@@ -44,50 +44,48 @@ void HX711::set_gain(byte gain) {
...
+    // Data: |--------|--------|--------|--------|
+    // Bit#:  33222222 22221111 11111100 00000000
+    //        10987654 32109876 54321098 76543210
+    union DataBuffer {
+        byte data[4];
+        long value;
+    } data_buffer;
+
+    // Wait for the chip to become ready
+    for (; !is_ready() ;) {
```
Comment (ReDetection) [inline HX711.cpp]: Why do you think `for` is better than `while`? Was it so necessary to change?

File: HX711.cpp (hunk 2 above)
Comment (zfields) [inline HX711.cpp]: The comment did not change, the capitalization now matches the rest.

File: HX711.cpp (hunk 4 above)
Comment (zfields) [inline HX711.cpp]: `for` is an optimization over `while` in several languages and is habit, otherwise no particular reason on Arduino.

File: HX711.cpp (hunk 1 above)
Comment (zfields) [inline HX711.cpp]: would have been by accident, I have too many laptops and IDEs

**Comment by zfields** [comment]
Sorry for any heartache here. I wrote the change when I was working with the hardware and tried to fix the merge conflict after I no longer had access to the hardware. That is why I requested it be *tested* then pulled.

File: HX711.cpp (hunk 1 above)
Comment (ReDetection) [inline HX711.cpp]: That's why github asks to review changes before submit. And it's a good practice to review staged changes before commit.

File: HX711.cpp (hunk 4 above)
Comment (ReDetection) [inline HX711.cpp]: It is nonsense. Arduino uses certain language and compiler is clever enough to optimise even more than you think. Don't waste readability for such purposes until it really needed.

File: HX711.cpp (hunk 2 above)
Comment (ReDetection) [inline HX711.cpp]: Then it should be a different pull request.

File: HX711.cpp (hunk 3 above)
Comment (ReDetection) [inline HX711.cpp]: This is premature optimisation. Compiler is clever enough to see you don't use iteration variables inside loop body and will find best loop variant for you.

File: HX711.cpp (hunk 1 above)
Comment (zfields) [inline HX711.cpp]: @ReDetection Flame off brother, put your energy in your code.
